### PR TITLE
*: refactor `context.Context`

### PR DESCRIPTION
**Pull request by coocood**
Make transaction life cycle under control, prepare for correctly handle schema out of date error.
The old `GetTxn(forceNew boll)` is replaced by `Txn` and `NewTxn`,
`CommitTxn` and `RollbackTxn` is removed.

### CODE REVIEW

File: executor/executor.go [hunk 1]
```diff
@@ -1390,6 +1384,7 @@ func (e *TableScanExec) Next() (*Row, error) {
 func (e *TableScanExec) nextForInfoSchema() (*Row, error) {
 	if e.infoSchemaRows == nil {
 		columns := make([]*table.Column, len(e.schema))
+		fmt.Printf("columns count %d, schema count %d\n", len(columns), len(e.schema))
```
Comment (tiancaiamao) [inline executor/executor.go]: log or remove ...

File: executor/executor_ddl.go [hunk 2]
```diff
@@ -72,6 +72,11 @@ func (e *DDLExec) Next() (*Row, error) {
 	if err != nil {
 		return nil, errors.Trace(err)
 	}
+	// Update InfoSchema in TxnCtx, so it will pass schema check.
```
Comment (tiancaiamao) [inline executor/executor_ddl.go]: add a TODO ?
schema check should not check schema change, and schema renew should not be automatically when schema check fail.

File: session.go [hunk 3]
```diff
@@ -313,32 +305,32 @@ func (s *session) Retry() error {
...
 				}
-				log.Warnf("session:%v, err:%v", s, err)
-				return errors.Trace(err)
 			}
 		}
-		if success {
+		if err == nil {
 			err = s.doCommit()
-			if !s.isRetryableError(err) {
+			if err == nil {
```
Comment (hanfei1991) [inline session.go]: how about if err = s.doCommit(); err == nil {...}

File: session.go [hunk 4]
```diff
@@ -313,32 +305,32 @@ func (s *session) Retry() error {
...
-			txt := st.OriginText()
-			if len(txt) > sqlLogMaxLen {
-				txt = txt[:sqlLogMaxLen]
-			}
-			log.Warnf("Retry %s (len:%d)", txt, len(st.OriginText()))
-			_, err = st.Exec(s)
-			if err != nil {
-				if s.isRetryableError(err) {
-					success = false
+		err = PrepareTxnCtx(s)
```
Comment (hanfei1991) [inline session.go]: how about if err = PrepareTxnCtx(s); err == nil {...}

File: ddl/column_test.go [hunk 5]
```diff
@@ -439,8 +446,12 @@ func (s *testColumnSuite) checkWriteOnlyColumn(c *C, ctx context.Context, d *ddl
 	if err != nil {
 		return errors.Trace(err)
 	}
+	err = ctx.Txn().Commit()
+	if err != nil {
+		return errors.Trace(err)
+	}
 
-	_, err = ctx.GetTxn(true)
```
Comment (zimulala) [inline ddl/column_test.go]: Use `NewTxn` is enough.

File: executor/executor_ddl.go (hunk 2 above)
Comment (coocood) [inline executor/executor_ddl.go]: After DDL execution, the session is not in transaction, so it's safe to update schema version.

File: session.go (hunk 4 above)
Comment (coocood) [inline session.go]: I think put an important statement in if condition obscures it, harder to read.

File: session.go (hunk 3 above)
Comment (coocood) [inline session.go]: ditto

**Comment by hanfei1991** [comment]
LGTM

File: plan/physical_plan_builder.go [hunk 6]
```diff
@@ -141,12 +141,8 @@ func (p *DataSource) convert2TableScan(prop *requiredProperty) (*physicalPlanInf
...
-	txn, err := p.ctx.GetTxn(false)
-	if err != nil {
-		return nil, errors.Trace(err)
-	}
-	if txn != nil {
-		ts.readOnly = txn.IsReadOnly()
+	if p.ctx.Txn() != nil {
+		ts.readOnly = p.ctx.Txn().IsReadOnly()
 	} else {
 		ts.readOnly = true
```
Comment (tiancaiamao) [inline plan/physical_plan_builder.go]: Would `p.ctx.Txn()` be nil here? and why set `ts.readOnly = true` when `p.ctx.Txn()` is nil ?

File: plan/physical_plan_builder.go [hunk 7]
```diff
@@ -222,12 +219,8 @@ func (p *DataSource) convert2IndexScan(prop *requiredProperty, index *model.Inde
...
-	txn, err := p.ctx.GetTxn(false)
-	if err != nil {
-		return nil, errors.Trace(err)
-	}
-	if txn != nil {
-		is.readOnly = txn.IsReadOnly()
+	if p.ctx.Txn() != nil {
+		is.readOnly = p.ctx.Txn().IsReadOnly()
 	} else {
 		is.readOnly = true
```
Comment (tiancaiamao) [inline plan/physical_plan_builder.go]: ditto

File: plan/physical_plan_builder.go (hunk 6 above)
Comment (coocood) [inline plan/physical_plan_builder.go]: Here ReadOnly means transaction has not written any data.
If txn is nil, there is no dirty data in the transaction, so it is read only.

Comment (coocood) [inline plan/physical_plan_builder.go]: Txn is nil in many tests, so check is needed.

**Comment by tiancaiamao** [comment]
LGTM

**Pull request merged**